### ISSUE TITLE
[CLIFF] Question on constants used for bbox_info and evaluation code

### ISSUE THREAD

**Issue by pangyyyyy** [huawei-noah/noah-research#94]
Dear authors, thanks for the great work and for providing the trained models! 

I have a few queries.

(1) Would you mind releasing your evaluation code for the CLIFF models on 3DPW-test set? 
(2) I am confused about some of the constants used in deriving the `bbox_info` i.e. 2.8,. 0.24, 0.06. What do these values mean? 
https://github.com/huawei-noah/noah-research/blob/7956f49b22dd7b5f1b18f660766342b9f68c9d92/CLIFF/demo.py#L124-L127

Thank you!

**Comment by zhihaolee** [comment]
Thanks for your interest!

1) We use the same evaluation code as SPIN (https://github.com/nkolot/SPIN/blob/master/eval.py), except the network input. You can modify it referring to our demo code, and get the same results.

2) These constants are used to normalize `bbox_info` to [-1, 1], and they are the statistical numbers from the Human3.6M dataset. However, I don't think they are the key factors.

**Comment by pangyyyyy** [comment]
Thanks! 

I modified SPIN's `eval.py` and the hmr model following the `demo.py` that you provide.

This is the code:
```
    for step, batch in enumerate(tqdm(data_loader, desc='Eval', total=len(data_loader))):
        # Get ground truth annotations from the batch
        gt_pose = batch['pose'].to(device)
        gt_betas = batch['betas'].to(device)
        gt_vertices = smpl_neutral(betas=gt_betas, body_pose=gt_pose[:, 3:], global_orient=gt_pose[:, :3]).vertices
        images = batch['img'].to(device)
        gender = batch['gender'].to(device)
        curr_batch_size = images.shape[0]

        # i saved out center, scale, img_h, img_w, focal_length to get bbox_info
        center = batch["center"].to(device).float()
        scale = batch["scale"].to(device).float()
        img_h = batch["img_h"].to(device).float()
        img_w = batch["img_w"].to(device).float()
        focal_length = batch["focal_length"].to(device).float()

        cx, cy, b = center[:, 0], center[:, 1], scale * 200
        bbox_info = torch.stack([cx - img_w / 2., cy - img_h / 2., b], dim=-1)
        bbox_info[:, :2] = bbox_info[:, :2] / focal_length.unsqueeze(-1) * 2.8  # [-1, 1]
        bbox_info[:, 2] = (bbox_info[:, 2] - 0.24 * focal_length) / (0.06 * focal_length)  # [-1, 1]
        full_img_shape = torch.stack((img_h, img_w), dim=-1)
        # print(bbox_info.shape)

        with torch.no_grad():
            pred_rotmat, pred_betas, pred_cam_crop = model(images, bbox_info)
            
            pred_cam_full = cam_crop2full(pred_cam_crop, center, scale, full_img_shape, focal_length)
            pred_output = smpl_neutral(betas=pred_betas,
                                    body_pose=pred_rotmat[:, 1:],
                                    global_orient=pred_rotmat[:, [0]],
                                    pose2rot=False,
                                    transl=pred_cam_full)
            pred_vertices = pred_output.vertices
```

However, using the provided `res50-PA45.7_MJE72.0_MVE85.3_3dpw.pt` file I obtained PA-MPJPE of 51.4. I was wondering if the above implementation is what you've done?

Hope to get your advice on this, thanks!

**Comment by zhihaolee** [comment]
I think the main reason may be the input image resolution. 
Please note that the resolution for these two released models is 256 * 192. 
If trained and evaluated on the same resolution, there is no much difference between 224 * 244 and 256 * 192.
Buf for the `res50-PA45.7_MJE72.0_MVE85.3_3dpw.pt` which is trained on 256 * 192, if fed with 224 * 224 input, the PA-MPJPE is 53.1 on my side.

**Comment by pangyyyyy** [comment]
Hi @zhihaolee, thanks for the clarification! after fixing the image resolution, I managed to obtain a PA-MPJPE of 46.7 , which is now much closer to the reported value.

Results from SPIN eval pipeline:
```
MPJPE: 74.47448722288111
Reconstruction Error: 46.68791574155502
```

Thanks!

**Comment by zhihaolee** [comment]
Glad to know that. 
The remaining difference could be the `center` and `scale` in the `3dpw_test.npz`. 
I projected the 3D SMPL joints to the image plane to re-generate the grount-truth 2D keypoints, and then `center` and `scale` using this scale calculation:
https://github.com/huawei-noah/noah-research/blob/7956f49b22dd7b5f1b18f660766342b9f68c9d92/CLIFF/common/imutils.py#L99

For the additional queries:
- Yes, 256x192 for the provided checkpoints.
- Yes, and the partition is [0.4, 0.3, 0.1, 0.2] for [H36M, in-the-wild datasets, 3DHP, 3DPW] following EFT.

**Comment by pangyyyyy** [comment]
@zhihaolee Thanks for the clarification!

**Comment by Baolar** [comment]
> Glad to know that. The remaining difference could be the `center` and `scale` in the `3dpw_test.npz`. I projected the 3D SMPL joints to the image plane to re-generate the grount-truth 2D keypoints, and then `center` and `scale` using this scale calculation:
> 
> https://github.com/huawei-noah/noah-research/blob/7956f49b22dd7b5f1b18f660766342b9f68c9d92/CLIFF/common/imutils.py#L99
> 
> For the additional queries:
> 
> * Yes, 256x192 for the provided checkpoints.
> * Yes, and the partition is [0.4, 0.3, 0.1, 0.2] for [H36M, in-the-wild datasets, 3DHP, 3DPW] following EFT.

@zhihaolee  How did you project 3D SMPL joints to the image plane for re-generating  2d ground truth? I have checked the 3DPW dataset. Did you use cam_instrincs, root_orientation, and 3D SMPL joints in 3DPW test labels to generate 2D keypoints and then use it to obtain center and bbox instead of OpenPose?

**Comment by zhihaolee** [comment]
Yes, and the cam_extrinsics are also needed.

**Comment by MooreManor** [comment]
@zhihaolee Regenerating all the `.npz` is a little bit complex for all the datasets. Could you provide the preprocessing scripts for re-generating `center` and `scale` in `.npz` of all training and testing datasets?

**Comment by MooreManor** [comment]
@zhihaolee 
Do you mean that only 3DPW dataset needs projected 2D keypoints? That is, for all the other datasets (e.g. h36m),  just use their 2d joints in the annotations to regenerate the bbox, and then use `bbox_from_detector` to create the new `center` and `scale`. 
https://github.com/huawei-noah/noah-research/blob/7956f49b22dd7b5f1b18f660766342b9f68c9d92/CLIFF/common/imutils.py#L86-L103

**Comment by zhihaolee** [comment]
@MooreManor 
Yes, you are right.

**Comment by MooreManor** [comment]
@zhihaolee 
I used [BOA script](https://github.com/syguan96/BOA/blob/20973879126145c0955503b4a3024bae13445482/utils/data_preprocess/pw3d_all.py#L95-L100) to regenrate the bbox, and then use your [bbox_from_detector](https://github.com/huawei-noah/noah-research/blob/7956f49b22dd7b5f1b18f660766342b9f68c9d92/CLIFF/common/imutils.py#L86-L103) to create `center` and `scale` for `3dpw_test.npz`. After  that, I used `hr48-PA43.0_MJE69.0_MVE81.2_3dpw.pt` to tested on `3dpw_test.npz`by using [SPIN eval.py](https://github.com/nkolot/SPIN/blob/master/eval.py). I have changed the testing part as [this](https://github.com/huawei-noah/noah-research/issues/94#issuecomment-1204003760), and for [BaseDataset](https://github.com/nkolot/SPIN/blob/master/datasets/base_dataset.py), I have also changed according to demo.py.

I did two evaluations on `3dpw_test.npz`, and got 71.10 and 70.06 MPJPE, which corresponded to `rescale` factor in `bbox_from_detector` 1.1 and 1.2 respectively. Did I miss something? Is it possible to send me the `3dpw_test.npz` by e-mail to check the difference?

**Comment by zhihaolee** [comment]
The rescale factor is 1.2, and the last difference could be the image resolution. I resize all the 3DPW images to half of their original size to reduce the image loading time.

**Comment by MooreManor** [comment]
@zhihaolee 
I also tested on h36-p2, but the result MPJPE is 57 using `hr48-PA43.0_MJE69.0_MVE81.2_3dpw.pt`, which is 10 scores higher than table 1. The rescale factor which I used in h36m is 0.9, and I followed the same procedure as on 3dpw. The score difference is much bigger than in the previous testing on 3dpw, so I want to confirm some details.

- What's the value of the rescale factor in your h36m testing?  
- Did you do any special operations for h36m pictures?

**Comment by zhihaolee** [comment]
1. The rescale factor is also 1.2 in the h36m testing.
2. No, just 3dpw.
3. We used h36m 3D joints of SMPL as GT for evaluation.
4. Both are 1.2 for MPII and COCO, too.

**Comment by MooreManor** [comment]
@zhihaolee 

I have two more questions about training.
- While training, do you use 2D/3D joints of GT SMPL or GT 2D/3D joints as supervision for 3d datasets (i.e. MPII_3D and h36m)? 
- What about that for 2d datasets (i.e. MPII and COCO) in training? Do you use pseudo 2D/3D joints of GT SMPL as supervision?

**Comment by zhihaolee** [comment]
@MooreManor 
- I haven't done that. Maybe you could try that, and tell us the results.
- We use the 2D/3D joints of GT SMPL ONLY for 3DPW, but GT 2D/3D joints for all the other datasets.

**Comment by ltphuongunited** [comment]
@pangyyyyy Can you send me your evaluation code for the CLIFF model on 3DPW-test set? I have tried but no success. Thank you so much.

**Comment by pangyyyyy** [comment]
Hi @ltphuongunited You can find the evaluation code under the 'cliff-eval' branch [here](https://github.com/pangyyyyy/spin/compare/master...cliff-eval) modified from SPIN's code. I got 46.68 for the ResNet-50 checkpoint (45.7 reported in paper), hope this helps!

**Comment by ltphuongunited** [comment]
Thanks @pangyyyyy so much, I got the same results as you (MPJPE: 74.74 and Reconstruction Error: 46.68).

**Comment by wulalah** [comment]
@MooreManor halo，i'm sorry to bother you.I'm reproducing the result on 3dpw dataset.I have some difficulties as you before. I need the reprocessed dataset annotation files. Have you regeneted all the .npz files?  Or could you provide the preprocessing scripts for re-generating center and scale in .npz of all training and testing datasets? I will thank you very much if you help me.
